### PR TITLE
support for file api reading

### PR DESCRIPTION
**Pull request by bdeng-xt**
This PR is an initial version for add reading support for file api. It added export_file_to_path helper function in ArrayUtil class

### CODE REVIEW

**Review by Shelnutt2 — APPROVED** [review]
Great first pass! Very fast addition thanks!

I'll leave it to you and @ihnorton on the follow-up items:
1) use vfs instead of ofstream for flexibility
2) take a parameter for size of buffer instead of allocating a vector for the full size up front

1) I only found a protected function xsputn from VFSFilebuf class to append bytes to a file, could you give me an example of using vfs to append
2) added a buffer_size as parameter

**Comment by Shelnutt2** [comment]
Xref: @snagles

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (Shelnutt2) [inline cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:35]: You can leave this as a char*, it doesn't need any size. The pointer is set to the Array owned value.

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (Shelnutt2) [inline cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:37]: This should be `output_file = std::string(original_name, name_size)`

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (Shelnutt2) [inline cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:65]: You don't need to set the subarray each time, calling `submit` again will resume the query where it left off.

**Review by Shelnutt2 — CHANGES_REQUESTED** [review]

**Comment by bdeng-xt** [comment]
> Shouldn't this be `Query::Status::COMPLETE`?

The Status of Query is wrapped as QueryStatus

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (bdeng-xt) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:65]: Thanks for your info. I was not certain about that yesterday. Will correct it soon

**Comment by ihnorton** [comment]
> The Status of Query is wrapped as QueryStatus

But we're using the C++ API here?

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (bdeng-xt) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc]: corrected it. thanks

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (bdeng-xt) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc]: corrected.

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (bdeng-xt) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc]: removed those codes to set the subarray. thanks

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (ihnorton) [inline cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:46]: Need to check for error here in case it is not writable

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (ihnorton) [inline cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:55]: This can be initialized with `buffer_size` as `data_buffer(buffer_size)`.

(In the future we should use a dedicated buffer class which does not initialize write zeros to the elements, because there's not need and it is expensive)

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (ihnorton) [inline cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:59]: This can give cause an infinite loop if the read never completes. We should have a counter here that checks the number of iterations without any results, and break out of the loop after something like 10 tries.

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (ihnorton) [inline cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:63]: eg
```
if (read_size == 0 && ++loop_zero_num > 10) { break; } else { read_size = 0; }
```

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (ihnorton) [inline cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:72]: We should catch `TileDBError` here and differentiate the error message.

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (ihnorton) [inline cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:138]: We need to check the status here, because `finalize` resets the status.

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (ihnorton) [inline cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:154]: Check `TileDBError`

**Review by ihnorton — APPROVED** [review]
Looking good, some comments inline. Also needs a test.

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (Shelnutt2) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:138]: Read queries should not call finalize

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (bdeng-xt) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc]: added check if the return is nullptr or not.

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (bdeng-xt) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc]: changed to data_buffer(buffer_size)

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (bdeng-xt) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:64]: added loop_zero_num as count variable

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (bdeng-xt) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc]: changed, thanks

Comment (ihnorton) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc]: (updated to use correct variable)

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (ihnorton) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc]: thanks

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (ihnorton) [inline cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:76]: Sorry I made a mistake in my example, this should be `loop_zero_num = 0`

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (bdeng-xt) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:87]: added

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (bdeng-xt) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:173]: added

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (bdeng-xt) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:153]: Yes, this is from another function save_file_from_path. I tested with a very simple png image file(Sample.png, about 1Mb). First, use save_file_from_path to save the png file to a tiledb array., then use export_file_to_path to export from the tiledb array to another png file(Sample_exported.png). The size and content of the original file and exported file are the same.

Comment (ihnorton) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:153]: We need to check `query_stats` after the write, _before_ calling `finalize`.

**Comment by bdeng-xt** [comment]
> For @Shelnutt2's requests I think these are done?
> 
> * [x]  use vfs instead of ofstream for flexibility
We don't have vfs_file_handle in cpp-api. so I can only use VFS::VFSFilebuf to read and write. Also in VFS::VFSFilebuf class, xsgetn and xsputn are all protected functions. 

> * [x]  take a parameter for size of buffer instead of allocating a vector for the full size up front
already added a buffer_size parameter
 
> 
> Not sure about this -- @bdeng-xt please make sure to respond to questions as a separate comment, because no one is notified about edits.

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (ihnorton) [inline cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:40]: Follow-up item: we need to check whether this is empty/unset, just in case.

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (ihnorton) [inline cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:53]: Setting this subarray means we will only read at-most `buffer_size`, rather than the whole file -- we need to query for `file_size`. We handle reading in chunks in the `while` loop below.

Comment (ihnorton) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:53]: I'm not sure what the goal is with `offset` here, seems unnecessary/cruft?

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (bdeng-xt) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:40]: added has_metadata check

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (bdeng-xt) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc]: changed the subarray to {0,file_size-1} and removed offset

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (ihnorton) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:40]: We don't handle the empty and unavailable case, so if there is no file specified and no metadata key, we will probably try to write a file name of empty string.

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (ihnorton) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:75]: Bump.

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (ihnorton) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:152]: Bump

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (ihnorton) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc]: Thanks

**Comment by ihnorton** [comment]
I marked all of my issues except 2 resolved, not sure if @Shelnutt2 wants to re-review his comments.

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (bdeng-xt) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:40]: added check for empty file

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (bdeng-xt) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:75]: corrected. thanks

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (bdeng-xt) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:152]: Are we checking if query_status is INCOMPLETE and do a while loop to continue write?

Comment (ihnorton) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:152]: No looping here, this is the write path, but we need to check `query_status == COMPLETE` or tell the user that their write did not complete.

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (ihnorton) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:40]: Sorry, what I mean is: we don't handle the condition where `output_file.empty() && !array->has_metadata(METADATA_ORIGINAL_FILE_NAME` -- no output file and no metadata filename available. IMO it should return an error, not try to write empty string or some undefined string.

**Review by Shelnutt2 — APPROVED** [review]

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (bdeng-xt) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc]: thanks for the explanation.

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (bdeng-xt) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc]: added the output when the query is not complete

File: cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc
Comment (ihnorton) [inline cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:163]: Shouldn't this be an error? We don't want them to think a file is written correctly if it is not actually complete... cc @Shelnutt2

Comment (Shelnutt2) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:163]: Agreed but this API is only for testing, we do not expect end users to call the write functionality yet only read from C#, so we have flexibility.

Comment (bdeng-xt) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:163]: added return statement when the write query is not complete

Comment (ihnorton) [reply cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc:163]: Thanks

File: examples/ExampleFileApi.cs
Comment (ihnorton) [inline examples/ExampleFileApi.cs:54]: Please remove before merge.

**Review by ihnorton — APPROVED** [review]
Thanks for addressing all the comments, @bdeng-xt 💯

File: examples/ExampleFileApi.cs
Comment (ihnorton) [reply examples/ExampleFileApi.cs:54]: Or make it a generic example uri.

Comment (bdeng-xt) [reply examples/ExampleFileApi.cs]: removed, thanks

**Comment by Shelnutt2** [comment]
Great work @bdeng-xt !

**Pull request merged**
